feat: add abstraction layer for mobile error tracking RFC (#9)
moxystudio/rfcs-oss: rfcs/0009-abstraction-layer-for-mobile-error-tracking.md

- Start Date: 2020-08-13
- Project: [React Native with MOXY](https://github.com/moxystudio/react-native-with-moxy)
- Implementation PR:

# Summary

The feature is an abstraction layer to work with different mobile error tracking solutions. The idea here is to have a common API that will be mapped to the particular API of the chosen error tracking solution.

# Basic example

Capturing an error:

```js
import createErrorTracking from "@moxy/react-native-sentry";

const errorTracking = createErrorTracking({ enabled: true });

try {
  aFunctionThatMightFail();
} catch (err) {
  errorTracking.captureError(err);
}
```

# Motivation

Let's suppose we were using [Sentry](https://sentry.io) as our solution. So, we would be able to capture an error using Sentry's `captureException` method.

```js
try {
  aFunctionThatMightFail();
} catch (err) {
  Sentry.captureException(err);
}
```

> However we might decide to use another error tracking solution later, e.g. [Firebase Crashlytics](https://firebase.google.com/products/crashlytics). By doing so we would need to go through all our codebase and change all Sentry's methods to their Crashlytics' counterparts, which would be an unnecessary work.

Reviewer: ```suggestion
However we might decide to use another error tracking solution later, e.g. [Firebase Crashlytics](https://firebase.google.com/products/crashlytics). By doing so we would need to go through all our codebase and change all Sentry's methods to their Crashlytics' counterparts, which would be cumbersome and time consuming work.
```

```js
try {
  aFunctionThatMightFail();
} catch (err) {
  crashlytics().recordError(err);
}
```

What we would like to to instead is mapping our own API to the one from our current error tracking solution. Basically we would only need to import another module.

Another important motivation is to add the ability to toggle the error reporting whenever we want to. Unfortunately, some solutions don't have this feature built-in, so we will create a way to handle this inside the abstraction layer.

# Detailed design

Example for Sentry:

```js
import * as Sentry from "@sentry/react-native";

const captureError = Sentry.captureException;

// level should be: fatal, error, warning, info, or debug.
const addBreadcrumb = ({ scope, message, level = "info" }) =>
  Sentry.addBreadcrumb({
    category: scope,
    message,
    level,
  });

const setUserId = (level) =>
  Sentry.configureScope((scope) => scope.setLevel(level));

// You can set the severity of an event to one of five values: fatal, error, warning, info, and debug. error is the default, fatal is the most severe, and debug is the least severe.

const setLevel = (id) =>
  Sentry.configureScope((scope) => scope.setUser({ id }));

> const enable = (enabled) => {

Reviewer: We must reference `isTrackingEnabled` inside the factory function `createErrorTracking`. Otherwise, we cannot leverage the closure.

Author: Oops, sure, my bad. Nice catch!

  isTrackingEnabled = enabled;
};

const createErrorTracking = ({ enabled = false } = {}) => {
  let isTrackingEnabled = enabled;

  const createTrackingFn = (fn) => (...args) => {
    if (!isTrackingEnabled) {
      return Promise.resolve();
    }

    return fn(...args);
  };

  return {
    captureError: createTrackingFn(captureError),
    addBreadcrumb: createTrackingFn(addBreadcrumb),
    setUserId: createTrackingFn(setUserId),
    setLevel: createTrackingFn(setLevel),
    enable,
  };
};

export default createErrorTracking;
```

Example for Firebase Crashlytics:

```js
import crashlytics from "@react-native-firebase/crashlytics";

const captureError = crashlytics().recordError;

const setUserId = crashlytics().setUserId;

const setLevel = (level) => crashlytics().setAttributes({ level });

// level should be: fatal, error, warning, info, or debug.

> const addBreadcrumbs = ({ scope, message, level = "info" }) =>

Reviewer: ```suggestion
const addBreadcrumb = ({ scope, message, level = "info" }) =>
```

  crashlytics().setAttributes({ category: scope, message, level });

> const enable = (enabled) => {

Reviewer: We must reference `isTrackingEnabled` inside the factory function `createErrorTracking`. Otherwise, we cannot leverage the closure behavior.

  isTrackingEnabled = enabled;
};

const createErrorTracking = ({ enabled = false } = {}) => {
  let isTrackingEnabled = enabled;

  const createTrackingFn = (fn) => (...args) => {
    if (!isTrackingEnabled) {
      return Promise.resolve();
    }

    return fn(...args);
  };

  return {
    captureError: createTrackingFn(captureError),
    addBreadcrumbs: createTrackingFn(addBreadcrumbs),
    setLevel: createTrackingFn(setLevel),
    setUserId: createTrackingFn(setUserId),
    enable,
  };
};

export default createErrorTracking;
```

## Usage examples

```js
import createErrorTracking from "@moxy/react-native-sentry";

> const errorTracking = createErrorTracking({ isEnable: true });

Reviewer: ```suggestion
const errorTracking = createErrorTracking({ enabled: true });
```

try {
  aFunctionThatMightFail();
} catch (err) {
  errorTracking.captureError(err);
}
```

> ## Regarding the initialization

Reviewer: I don't see why we have to delegate initialization to the developer. From what I've learned, both Sentry and BugSnag use a plain object with several options, which is a very standard approach to handle configuration. Thus, I'd say we can simply accept such options in the factory function and forward them to the specific initialization function.

Author: I thought that Bugsnag was basically out of the game, since we realized that Crashlytics should suite small things, while Sentry would do for the bigger ones because it has a better pricing in comparison to Bugsnag. That said, Sentry and Crashlytics are completely different, as I've exposed here.

Reviewer: For the time being we're not planning to use BugSnag, but we never know... Regardless, with or without it, the same rationale applies. Crashlytics does not require any initialization on JS side? Great! Nothing has to be done. If it ever does, we can add it later. Sentry takes an object with a few options which I believe a reasonable approach we can adopt for whatever solution we may end up using.
If we move the initialization logic up to the caller, we have to import the SDK of the solution in the app in addition to importing in the implementation modules themselves, so we end up leaking implementation details because the app has to know which SDK to import.
If we effectively go down the road of performing initialization in the app, it would probably be better to declare the SDK as a peer dependency in the implementation modules, install it manually in the app and pass it as an argument to the error tracking factory function.

Reviewer: We decided to go for peer dependencies to have more control over the initialization process, which becomes a responsibility of the app, and the version of the SDK. The instance of the SDK is passed as an argument to the implementing module.

In my opinion the initialization should not be abstracted, since Sentry and Firebase Crashlytics have quite differente ways of being initialized.

- Sentry:

```js
import * as Sentry from '@sentry/react-native';

Sentry.init({
    dsn: // Sentry's DSN,
});
```

- Firebase Crashlytics:

```js
import crashlytics from "@react-native-firebase/crashlytics";

// you can already use it inside the file
```

# Drawbacks

The only drawback I can see is the initial effort to design the solution, since the maintenance should be really simple.

# Alternatives

The alternative would be using the API provided directly from the error tracking solutions. However, this is not a good idea as I've already commented about in the "Motivation" section.

# Adoption strategy

> If we implement this proposal, the developers would need to check our own module which is probably going to be called `@moxy/react-native-{solution}`, where `solution` would be the name of the chosen error tracking solution (it could be `sentry` for instance) and see the available methods. I will probably add this to the RNWM documentation, to make its usage as clear as possible.

Reviewer: The module themselves will have a README file so their API and usage will be documented. What we will do instead is to write down a recipe to walk developers through how to integrate these modules in the boilerplate.

Author: Ok, I see! I will rewrite this

Author: Just pushed a commit with the rewrite. Let me know if it's what is supposed to be this time, please.

Author: @acostalima could you check this, please?

Reviewer: Check my latest review [below](https://github.com/moxystudio/rfcs-oss/pull/9#pullrequestreview-471663711) please. 🙏

  return (
    <ErrorTrackingContext.Provider>{children}</ErrorTrackingContext.Provider>
  );
};

export const useErrorTracking = () => {
  const context = useContext(ErrorTrackingContext);

  if (!context) {
    throw new Error(
      "useErrorTracking should be used within an ErrorTrackingProvider"
    );
  }

  return context;
};
```

```js
// src/app/App.js
import * as Sentry from "@sentry/react-native";
import createErrorTracking from "@moxy/react-native-sentry";
import { ErrorTrackingProvider } from "../../shared/react-error-tracking";

Sentry.init({
  dsn: // Sentry's DSN,
});

const errorTracking = createErrorTracking(Sentry, { enabled: true });

const App = () => (
  <ErrorTrackingProvider value={errorTracking}>
    {/* ... */}
  </ErrorTrackingProvider>
);
```

## Usage examples

Suppose there is a `Button.js` component:

```js
import React, { useCallback } from "react";
import { TouchableOpacity, Text } from "react-native";
import { useErrorTracking } from "../react-error-tracking";

const aFunctionThatMightFail = () => {
  throw new Error("Failed");
};

const Button = () => {
  const errorTracking = useErrorTracking();

>   const onPress = useCallback(() => {

Reviewer 2: Just a suggestion, in these cases we could create a hook that uses `useCallback` and performs the try/catch with error tracking. This way we would only use a hook like we would use `useCallback` and the try/catch would be "automatic"

Author: Thanks for your suggestion @ruipneves ! Didn't think about this possibility to be honest. I will definitely keep that in mind once I start working on the implementation

      </TouchableOpacity>

      <Text>Error tracking is {enabled ? "enabled" : "disabled"}</Text>
    </View>
  );
};
```

## SDK initialization

In my opinion the initialization should not be abstracted, since Sentry and Firebase Crashlytics have quite differente ways of being initialized.

### Sentry

```js
import * as Sentry from '@sentry/react-native';

Sentry.init({
    dsn: // Sentry's DSN,
});
```

### Firebase Crashlytics

```js
import crashlytics from "@react-native-firebase/crashlytics";

// you can already use it inside the file
```

# Drawbacks

The only drawback I can see is the initial effort to design the solution, since the maintenance should be really simple.

# Alternatives

The alternative would be using the API provided directly from the error tracking solutions. However, this is not a good idea as I've already commented about in the "Motivation" section.

# Adoption strategy

To integrate those modules `@moxy/react-native-{solution}`, where `solution` would be the name of the chosen error tracking solution (it could be `sentry` for instance), one should start by installing it:

```sh
npm i @sentry/react-native @moxy/react-native-sentry

> # yarn add @sentry/react-native @moxy/react-native-sentry

Reviewer: Improved the text a bit:

"New and existing apps are required to install the SDK of the chosen error tracking solution and the respective module which implements the abstraction for such SDK:

```sh
npm i @sentry/react-native @moxy/react-native-{solution}

# yarn add @sentry/react-native @moxy/react-native-{solution}
```

where `solution` would be the name of the solution to use, such as `crashlytics` or `sentry`.
Existing apps would also have to be refactored to start using the API specified here."

Author: 👍

- Start Date: 2020-08-13
- Project: [React Native with MOXY](https://github.com/moxystudio/react-native-with-moxy)
- Implementation PR:

# Summary

The feature is an abstraction layer to work with different mobile error tracking solutions. The idea here is to have a common API that will be mapped to the particular API of the chosen error tracking solution.

# Basic example

Capturing an error:

```js
import createErrorTracking from "@moxy/react-native-sentry";

> const errorTracking = createErrorTracking({ enabled: true });

Reviewer: ```suggestion
import * as sentry from "@sentry/react-native";
import createErrorTracking from "@moxy/react-native-sentry";

const errorTracking = createErrorTracking(sentry, { enabled: true });
```

Right?

try {
  aFunctionThatMightFail();
} catch (err) {
  errorTracking.captureError(err);
}
```

# Motivation

Let's suppose we were using [Sentry](https://sentry.io) as our solution. So, we would be able to capture an error using Sentry's `captureException` method.

```js
try {
  aFunctionThatMightFail();
} catch (err) {
  Sentry.captureException(err);
}
```

However we might decide to use another error tracking solution later, e.g. [Firebase Crashlytics](https://firebase.google.com/products/crashlytics). By doing so we would need to go through all our codebase and change all Sentry's methods to their Crashlytics' counterparts, which would becumbersome and time consuming work.

```js
try {
  aFunctionThatMightFail();
} catch (err) {
  crashlytics().recordError(err);
}
```

What we would like to to instead is mapping our own API to the one from our current error tracking solution. Basically we would only need to import another module.

Another important motivation is to add the ability to toggle the error reporting whenever we want to. Unfortunately, some solutions don't have this feature built-in, so we will create a way to handle this inside the abstraction layer.

# Detailed design

> Example for Sentry:

Reviewer: ```suggestion
## Implementation for Sentry
```

```js
const createErrorTracking = (sentry, { enabled = false } = {}) => {
  let isTrackingEnabled = enabled;

  const createTrackingFn = (fn) => (...args) => {
    if (!isTrackingEnabled) {
      return Promise.resolve();
    }

    return fn(...args);
  };

>   const captureError = sentry.captureException;

Reviewer: Beware this might not work. Sentry may use `this` internally, so we have to bind `captureException` to Sentry's instance:

```js
const captureError = sentry.captureException.bind(sentry);
```

I'd rather keep things simple and use the following instead:

```js
const captureError = (error) => sentry.captureException(error);
```

The final implementation could be:

```js
{
...
captureError: createTrackingFn((error) => sentry.captureException(error)),
...
}
```

Author: Good call, didn't think about this. Thanks :raised_hands:

  // level should be: fatal, error, warning, info, or debug.
  const addBreadcrumb = ({ scope, message, level = "info" }) =>
    sentry.addBreadcrumb({
      category: scope,
      message,
      level,
    });

  const setUserId = (level) =>
    sentry.configureScope((scope) => scope.setLevel(level));

  // You can set the severity of an event to one of five values: fatal, error, warning, info, and debug. error is the default, fatal is the most severe, and debug is the least severe.

  const setLevel = (id) =>
    sentry.configureScope((scope) => scope.setUser({ id }));

  const enable = (enabled) => {
    isTrackingEnabled = enabled;
  };

  return {
    captureError: createTrackingFn(captureError),
    addBreadcrumb: createTrackingFn(addBreadcrumb),
    setUserId: createTrackingFn(setUserId),
    setLevel: createTrackingFn(setLevel),
    enable,
  };
};

export default createErrorTracking;
```

> Example for Firebase Crashlytics:

Reviewer: ```suggestion
## Implementation for Firebase Crashlytics
```

```js
const createErrorTracking = (crashlytics, { enabled = false } = {}) => {

>   let isTrackingEnabled = enabled;

Reviewer: I believe we won't need dedicated logic to keep track of whether Crashlytics is enabled or disabled. This is built-in into the SDK, right?

Author: Crashlytics provides the `setCrashlyticsCollectionEnabled` method that does it. However we would need to change the public API to use it. You can take a look at an example here: https://rnfirebase.io/crashlytics/usage#opt-out

Reviewer: Why would we need to change the API?

Author: Sorry, I was wrong. We don't need to change the API, only the implementation. What do you think about this:

```js
const createErrorTracking = (crashlytics, { enabled = false } = {}) => {
  const enable = async (enabled) => {
    await crashlytics.setCrashlyticsCollectionEnabled(enabled);
  };

  enable(enabled);

  const captureError = crashlytics.recordError;

  const setUserId = crashlytics.setUserId;

  const setLevel = (level) => crashlytics.setAttributes({ level });

  // level should be: fatal, error, warning, info, or debug.
  const addBreadcrumb = ({ scope, message, level = "info" }) =>
    crashlytics.setAttributes({ category: scope, message, level });

  return {
    captureError,
    addBreadcrumb,
    setLevel,
    setUserId,
    enable,
  };
};

export default createErrorTracking;
```

Reviewer: Yeah, I think we go after something like that. There is no need to use async/await over there though.

Author: `setCrashlyticsCollectionEnabled` returns a promise https://rnfirebase.io/reference/crashlytics#setCrashlyticsCollectionEnabled

Reviewer: I know, but you can just return the promise. There is no need to wrap it with async/await.

Author: Then the usage of this module would be different than the usage of Sentry's one. The Sentry's module `enable` function doesn't return a promise

Author: Ok, after we talked, I realized that it makes sense to turn the `enable` function from the Sentry's module into an async one

  const createTrackingFn = (fn) => (...args) => {
    if (!isTrackingEnabled) {
      return Promise.resolve();
    }

    return fn(...args);
  };

>   const captureError = crashlytics().recordError;

Reviewer: I think it's reasonable to assume Crashlytics' instance will be provided by the caller:

```suggestion
  const captureError = crashlytics.recordError;
```

Also, beware this approach might not work too because `recordError` is not bound to `crashlytics`.

>   const setUserId = crashlytics().setUserId;

Reviewer: I think it's reasonable to assume Crashlytics' instance will be provided by the caller:

```suggestion
  const captureError = crashlytics.setUserId;
```

Also, beware this approach might not work too because `setUserId` is not bound to `crashlytics`.

>   const setLevel = (level) => crashlytics().setAttributes({ level });

Reviewer: ```suggestion
  const setLevel = (level) => crashlytics.setAttributes({ level });
```

  // level should be: fatal, error, warning, info, or debug.
  const addBreadcrumb = ({ scope, message, level = "info" }) =>

>     crashlytics().setAttributes({ category: scope, message, level });

Reviewer: ```suggestion
    crashlytics.setAttributes({ category: scope, message, level });
```

  const enable = (enabled) => {
    isTrackingEnabled = enabled;
  };

  return {
    captureError: createTrackingFn(captureError),
    addBreadcrumb: createTrackingFn(addBreadcrumb),
    setLevel: createTrackingFn(setLevel),
    setUserId: createTrackingFn(setUserId),
    enable,
  };
};

export default createErrorTracking;
```

## Setup

```js

> // src/shared/hooks/error-tracking.js

Reviewer: ```suggestion
// src/shared/react-error-tracking.js
```

import React, { createContext, useContext } from "react";

const ErrorTrackingContext = createContext();

export const ErrorTrackingProvider = ({ children }) => {
  return (
    <ErrorTrackingContext.Provider>{children}</ErrorTrackingContext.Provider>
  );
};

export const useErrorTracking = () => {
  const context = useContext(ErrorTrackingContext);

  if (!context) {
    throw new Error(
      "useErrorTracking should be used within an ErrorTrackingProvider"
    );
  }

  return context;
};
```

```js
// src/app/App.js
import * as Sentry from "@sentry/react-native";
import createErrorTracking from "@moxy/react-native-sentry";

> import { ErrorTrackingProvider } from "../../shared/hooks/error-tracking";

Reviewer: ```suggestion
import { ErrorTrackingProvider } from "../../shared/react-error-tracking";
```

> import { ErrorTrackingProvider } from "../../shared/hooks/error-tracking";

Reviewer: ```suggestion
import { ErrorTrackingProvider } from "../../shared/react-error-tracking";
```

Sentry.init({
  dsn: // Sentry's DSN,
});

const errorTracking = createErrorTracking(Sentry, { enabled: true });

const App = () => (
  <ErrorTrackingProvider value={errorTracking}>
    {/* ... */}
  </ErrorTrackingProvider>
);
```

## Usage examples

Suppose there is a `Button.js` component:

```js
import React, { useCallback } from "react";
import { TouchableOpacity, Text } from "react-native";

> import { useErrorTracking } from "../hooks/error-tracking";

Reviewer: ```suggestion
import { useErrorTracking } from "../react-error-tracking";
```

> const errorTracking = useErrorTracking();

Reviewer: Hooks can only be used within render functions.

Author: Oops, sure

const aFunctionThatMightFail = () => {
  throw new Error("Failed");
};

const Button = () => {
  const onPress = useCallback(() => {
    try {
      aFunctionThatMightFail();
    } catch (err) {
      errorTracking.captureError(err);
    }
  }, []);

  return (
    <TouchableOpacity onPress={onPress}>
      <Text>Press Here</Text>
    </TouchableOpacity>
  );
};
```

Note that it's possible to enable or disable the error tracking as you wish, so it would be possible to:

```js
import React, { useState, useCallback } from "react";

> import { View, ouchableOpacity, Text } from "react-native";

Reviewer: ```suggestion
import { View, TouchableOpacity, Text } from "react-native";
```

> import createErrorTracking from "@moxy/react-native-sentry";

Reviewer: ```suggestion
```

> import { useErrorTracking } from "../hooks/error-tracking";

Reviewer: ```suggestion
import { useErrorTracking } from "../react-error-tracking";
```

const ToggleErrorTracking = () => {

>   const [enabled, setEnabled] = useState(true);
> 
>   const errorTracking = useErrorTracking();

Reviewer: ```suggestion
  const [enabled, setEnabled] = useState(true);
  const errorTracking = useErrorTracking();
```

Author: Do you really want me to just remove a line here?

Reviewer: No. Please ignore my suggestion.

  const onPress = useCallback(() => {
    setEnabled((state) => !state);

    errorTracking.enable(enabled);
  }, [setEnabled, errorTracking]);

  return (
    <View>
      <TouchableOpacity onPress={onPress}>
        <Text>Toggle Error Tracking</Text>
      </TouchableOpacity>

      <Text>Error tracking is {enabled ? "enabled" : "disabled"}</Text>
    </View>
  );
};
```

> ## Regarding the initialization

Reviewer: ```suggestion
## SDK initialization
```

In my opinion the initialization should not be abstracted, since Sentry and Firebase Crashlytics have quite differente ways of being initialized.

> - Sentry:

Reviewer: ```suggestion
### Sentry
```

```js
import * as Sentry from '@sentry/react-native';

Sentry.init({
    dsn: // Sentry's DSN,
});
```

> - Firebase Crashlytics:

Reviewer: ```suggestion
### Firebase Crashlytics
```

```js
import crashlytics from "@react-native-firebase/crashlytics";

// you can already use it inside the file
```

# Drawbacks

The only drawback I can see is the initial effort to design the solution, since the maintenance should be really simple.

# Alternatives

The alternative would be using the API provided directly from the error tracking solutions. However, this is not a good idea as I've already commented about in the "Motivation" section.

# Adoption strategy

To integrate those modules `@moxy/react-native-{solution}`, where `solution` would be the name of the chosen error tracking solution (it could be `sentry` for instance), one should start by installing it:

```sh

> npm i @moxy/react-native-sentry

Reviewer: ```suggestion
npm i @sentry/react-native @moxy/react-native-sentry
```

> # yarn add @moxy/react-native-sentry

Reviewer: ```suggestion
# yarn add @sentry/react-native @moxy/react-native-sentry
```

Author: Nice catch 👍

> ```
> 
> On the `src/app/App.js` file, you should initialize Sentry:
> 
> ```js
> // src/app/App.js
> import * as Sentry from "@sentry/react-native";
> import createErrorTracking from "@moxy/react-native-sentry";
> import { ErrorTrackingProvider } from "../../shared/hooks/error-tracking";
> 
> Sentry.init({
>   dsn: // Sentry's DSN,
> });
> 
> const errorTracking = createErrorTracking(Sentry, { enabled: true });
> 
> const App = () => (
>   <ErrorTrackingProvider value={errorTracking}>
>     {/* ... */}
>   </ErrorTrackingProvider>
> );
> ```

Reviewer: This is already included in the setup section so there is no need to repeat it here.

```suggestion
```

Author: I don't see repeating this here as problem to be honest, since this is just there to make it clear how to the initialization should be done, but I will remove it anyways.
